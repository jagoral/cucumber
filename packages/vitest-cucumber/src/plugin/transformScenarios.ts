import * as recast from "recast";

const { identifier, memberExpression, callExpression } = recast.types.builders;
const { MemberExpression, ExpressionStatement, Identifier, CallExpression } =
  recast.types.namedTypes;

const scenarioNames = ["scenario", "scenarioOutline"];

export function transformScenarios(code: string): string {
  const ast = recast.parse(code);
  recast.visit(ast, {
    visitCallExpression(path) {
      const { node } = path;

      if (
        Identifier.check(node.callee) &&
        scenarioNames.includes(node.callee.name)
      ) {
        const rootCallExpressionPath = findRootCallExpressionPath(path);

        const rootCallExpressionNode = rootCallExpressionPath?.node;

        if (
          rootCallExpressionPath &&
          CallExpression.check(rootCallExpressionNode) &&
          !isBuildAlreadyAdded(rootCallExpressionNode)
        ) {
          rootCallExpressionPath.replace(
            callExpression(
              memberExpression(rootCallExpressionNode, identifier("build")),
              [],
            ),
          );
        }
      }

      this.traverse(path);
    },
    visitMemberExpression(path) {
      const { node } = path;

      // Check if the member expression is a scenario object
      // Case: scenario.only, scenario.skip
      if (
        Identifier.check(node.object) &&
        scenarioNames.includes(node.object.name)
      ) {
        const rootCallExpressionPath = findRootCallExpressionPath(path);

        const rootCallExpressionNode = rootCallExpressionPath?.node;

        if (
          rootCallExpressionPath &&
          CallExpression.check(rootCallExpressionNode) &&
          !isBuildAlreadyAdded(rootCallExpressionNode)
        ) {
          rootCallExpressionPath.replace(
            callExpression(
              memberExpression(rootCallExpressionNode, identifier("build")),
              [],
            ),
          );
        }
      }

      this.traverse(path);
    },
  });

  return recast.prettyPrint(ast).code.toString();
}

function findRootCallExpressionPath(path: any) {
  if (!path.parentPath) {
    return null;
  }

  const { parentPath } = path;
  const parentPathNode = parentPath.node;

  if (parentPathNode && ExpressionStatement.check(parentPathNode)) {
    return path;
  }

  return findRootCallExpressionPath(parentPath);
}

function isBuildAlreadyAdded(
  rootChainNode: recast.types.namedTypes.CallExpression,
): boolean {
  return (
    MemberExpression.check(rootChainNode.callee) &&
    Identifier.check(rootChainNode.callee.property) &&
    rootChainNode.callee.property.name === "build"
  );
}
