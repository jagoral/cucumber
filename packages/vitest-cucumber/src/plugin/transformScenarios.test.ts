import { format } from "prettier";
import { describe, expect, it } from "vitest";
import { transformScenarios } from "./transformScenarios";

describe("transformScenarios", () => {
  async function getTransformedScenario(code: string) {
    return format(transformScenarios(code), {
      parser: "typescript",
    });
  }

  it("should add .build() at the end of the chain", async () => {
    const code = `
      scenario("test")
        .when("a", () => {})
        .then("b", () => {})
        .and("c", () => {});
    `;

    expect(await getTransformedScenario(code)).toMatchSnapshot();
  });

  it("should add .build() at the end of the scenarioOutline chain", async () => {
    const code = `
      scenarioOutline("test")
        .when("a", () => {})
        .then("b", () => {})
        .and("c", () => {});
    `;

    expect(await getTransformedScenario(code)).toMatchSnapshot();
  });

  it("should skip adding .build() when its already added", async () => {
    const code = `
    scenario("test")
      .when("a", () => {})
      .then("b", () => {})
      .and("c", () => {})
      .build();
  `;

    expect(await getTransformedScenario(code)).toMatchSnapshot();
  });

  it("should add .build() to all scenarios", async () => {
    const code = `
    scenario("test1")
      .when("a", () => {})
      .then("b", () => {})
      .and("c", () => {});

    scenario("test2")
      .given("something", {})
      .when("a", () => {})
      .then("b", () => {});
  `;

    expect(await getTransformedScenario(code)).toMatchSnapshot();
  });

  it("should add .build() when chain identifier", async () => {
    const code = `
    scenario.only("test")
      .when("a", () => {})
      .then("b", () => {})
      .and("c", () => {});
  `;

    expect(await getTransformedScenario(code)).toMatchSnapshot();
  });
});
