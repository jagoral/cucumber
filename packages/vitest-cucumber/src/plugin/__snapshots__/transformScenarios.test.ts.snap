// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformScenarios > should add .build() at the end of the chain 1`] = `
"scenario("test")
  .when("a", () => {})
  .then("b", () => {})
  .and("c", () => {})
  .build();
"
`;

exports[`transformScenarios > should add .build() to all scenarios 1`] = `
"scenario("test1")
  .when("a", () => {})
  .then("b", () => {})
  .and("c", () => {})
  .build();
scenario("test2")
  .given("something", {})
  .when("a", () => {})
  .then("b", () => {})
  .build();
"
`;

exports[`transformScenarios > should add .build() when chain identifier 1`] = `
"scenario
  .only("test")
  .when("a", () => {})
  .then("b", () => {})
  .and("c", () => {})
  .build();
"
`;

exports[`transformScenarios > should skip adding .build() when its already added 1`] = `
"scenario("test")
  .when("a", () => {})
  .then("b", () => {})
  .and("c", () => {})
  .build();
"
`;
